### PR TITLE
Allow manually specifying aliases for contract methods

### PR DESCRIPTION
**Pull request by nlordell**
This PR implements manually specifying alias for solidity contract methods. The main motivation here is to allow for:
1. Overloaded methods
2. Deal with name collisions, between events a methods for example (`transfer`, I'm looking at you...)
3. Get code generation ready for overloading event names

### Test Plan

New unit tests and example.

### CODE REVIEW

File: derive/src/lib.rs [hunk 1]
```diff
@@ -84,7 +96,7 @@ pub fn contract(input: TokenStream) -> TokenStream {
 
     let span = args.span();
     expand(args.into_inner())
-        .unwrap_or_else(|e| SynError::new(span, e.to_string()).to_compile_error())
+        .unwrap_or_else(|e| SynError::new(span, format!("{:?}", e)).to_compile_error())
```
Comment (fleupold) [inline derive/src/lib.rs:101]: Is this change intended?

File: generate/src/contract.rs [hunk 2]
```diff
@@ -87,6 +89,17 @@ impl Context {
         };
         let contract_name = util::ident(raw_contract_name);
 
+        let mut method_aliases = HashMap::new();
+        for (signature, alias) in args.method_aliases.into_iter() {
+            let alias = syn::parse_str(&alias)?;
+            if method_aliases.insert(signature.clone(), alias).is_some() {
+                return Err(anyhow!(
```
Comment (fleupold) [inline generate/src/contract.rs:99]: Here you are not checking for duplicate alias?

File: src/contract.rs [hunk 3]
```diff
@@ -114,12 +129,19 @@ impl<T: Transport> Instance<T> {
...
     {
-        let function = self.abi.function(name.as_ref())?;
+        // NOTE: First try to get the method by signature, and if that fails
+        //   then by name.
+        let signature = signature.as_ref();
+        let function = self
+            .methods
+            .get(signature)
+            .map(|(name, index)| Ok(&self.abi.functions[name][*index]))
+            .unwrap_or_else(|| self.abi.function(signature))?;
```
Comment (fleupold) [inline src/contract.rs]: In which case is this possible? If I understand correctly we always pass in the signature literal now, so if it doesn't exist in the Map you created it won't exist at all, would it? `abi.function` seems to work on method names not signatures.

**Review by fleupold — APPROVED** [review]
Quite a handful to review here, it's nice to see how these features blend in nicely into the current architecture without much refactorings.

File: derive/src/lib.rs (hunk 1 above)
Comment (nlordell) [reply derive/src/lib.rs:101]: Yes, it makes compile errors easier to debug. I've opened a separate PR #239 as its technically unrelated.

File: generate/src/contract.rs (hunk 2 above)
Comment (nlordell) [reply generate/src/contract.rs:99]: No because we will get compile errors on duplicate aliases as there will be two methods with the same identifier.

File: src/contract.rs (hunk 3 above)
Comment (nlordell) [reply src/contract.rs]: This is just for backwards compatibility in case someone was using the old `Instance` API. Now that you brought it up, it may not be worth supporting it.

File: generate/src/contract.rs (hunk 2 above)
Comment (nlordell) [reply generate/src/contract.rs:99]: I'll add a note.

**Review by e00E — APPROVED** [review]

File: common/src/errors.rs [hunk 4]
```diff
@@ -50,3 +50,8 @@ pub enum LinkError {
     #[error("undefined library {0}")]
     UndefinedLibrary(String),
 }
+
+/// An error reprensenting an error parsing a parameter type.
```
Comment (bh2smith) [inline common/src/errors.rs:54]: representing*

File: derive/src/lib.rs [hunk 5]
```diff
@@ -16,7 +18,10 @@ use std::collections::HashSet;
 use std::error::Error;
 use syn::ext::IdentExt;
 use syn::parse::{Error as ParseError, Parse, ParseStream, Result as ParseResult};
-use syn::{braced, parse_macro_input, Error as SynError, Ident, LitInt, LitStr, Token, Visibility};
+use syn::{
+    braced, parenthesized, parse_macro_input, Error as SynError, Ident, LitInt, LitStr, Token,
+    Visibility,
+};
```
Comment (bh2smith) [inline derive/src/lib.rs:24]: I assume this is some sort of change to the linter's line width since there does not appear to be anything new imported here.

File: derive/src/lib.rs [hunk 6]
```diff
@@ -72,8 +81,11 @@ use syn::{braced, parse_macro_input, Error as SynError, Ident, LitInt, LitStr, T
 ///     mod = my_contract_instance,
 ///     contract = MyContractInstance,
 ///     deployments {
-///         4 => "0x000102030405060708090a0b0c0d0e0f10111213"
-///         5777 => "0x0123456789012345678901234567890123456789"
+///         4 => "0x000102030405060708090a0b0c0d0e0f10111213",
+///         5777 => "0x0123456789012345678901234567890123456789",
+///     },
+///     methods {
+///         myMethod(uint256,bool) as my_renamed_method;
```
Comment (bh2smith) [inline derive/src/lib.rs:88]: Safe to assume this is the "meat" of the the PR's intended goal/functionality?

File: examples/truffle/contracts/OverloadedMethods.sol [hunk 7]
```diff
@@ -0,0 +1,14 @@
+pragma solidity ^0.5.0;
+
+/**
+ * @dev Contract to illustract how to work around overloaded Solidity methods.
```
Comment (bh2smith) [inline examples/truffle/contracts/OverloadedMethods.sol:4]: "illustract" is this a pun or a typo?

File: generate/src/contract/methods.rs [hunk 8]
```diff
@@ -95,24 +105,12 @@ fn expand_function(cx: &Context, function: &Function) -> Result<TokenStream> {
...
-        0 => String::new(),
-        _ => {
-            let mut params = function.inputs.iter().map(|param| &param.kind);
-            let first = params.next().expect("at least one param").to_string();
-            params.fold(first, |acc, param| format!("{},{}", acc, param))
-        }
-    };
-    format!("{}({})", function.name, types)
-}
-
```
Comment (bh2smith) [inline generate/src/contract/methods.rs:115]: Any particular reason this is no longer needed?

**Review by bh2smith — COMMENTED** [review]
Very cool! Can't want to see this in action!

File: derive/src/lib.rs (hunk 5 above)
Comment (nlordell) [reply derive/src/lib.rs:24]: `Visibility` was added.

File: examples/truffle/contracts/OverloadedMethods.sol (hunk 7 above)
Comment (nlordell) [reply examples/truffle/contracts/OverloadedMethods.sol:4]: I want to say its a pun, but its actually a typo.

File: generate/src/contract/methods.rs (hunk 8 above)
Comment (nlordell) [reply generate/src/contract/methods.rs:115]: This was implemented twice in two separate places, here and in `ethcontract_common`. I just removed this implementation and used the shared one instead.

File: derive/src/lib.rs (hunk 5 above)
Comment (bh2smith) [reply derive/src/lib.rs:24]: Ahh yes, my viability isn't what it used to be!
